### DISCUSSION TITLE
Hot reloading settings.json

### DISCUSSION THREAD

**Discussion by revsuine** [excalith/excalith-start-page#75] [Q&A]
Hiya, firstly I'm not sure if this is a bug, I may just be doing something wrong, in which case I'd appreciate an explanation of how to do things correctly. I've done my best to read docs and search things up but I don't do web app dev so I don't really know what I'm doing other than following instructions.

Basically, I'm managing to get the startpage up at `localhost:3000`, but when I edit `settings.json`, the changes don't appear on the server, including when I restart the server with `pm2 restart` (also tried `pm2 reload` and `pm2 delete` followed by `pm2 start` again). I have also tried `yarn dev` and `yarn start` but those also seem to only use the config from before the server was started for the first time. I have also tried running `yarn build` again, to no avail.

I am able to type `config edit` in the startpage and paste my `settings.json` in there and it displays my settings as I would expect.

### To Reproduce

1. Clone `https://github.com/excalith/excalith-start-page.git` (or the ssh, I cloned the ssh)
2. Follow the [fork instructions on the wiki](https://github.com/excalith/excalith-start-page/wiki/Fork):
3. In the directory of the cloned repo, `rm -rf .github/workflows`
4. Install nodejs and yarn (for me, `sudo pacman -S nodejs yarn`)
5. In the directory of the repo, type `yarn` to install all dependencies
6. Change some things in `settings.json` *prior to starting the server*. For me, I changed the theme to `Catppuccin Mocha` and set my username.
7. Run `yarn dev` and go to `localhost:3000`. See that the webpage currently reflects the state of `settings.json`.
8. Ctrl C and then edit `settings.json` some more. For me, I changed the bookmarks to be the websites I like to use.
9. Run `yarn dev` or `yarn start` or `pm2 start "yarn start"` (or `pm2 start "yarn dev"`) and observe that the startpage appears as though `settings.json` was not changed since the first time you started the server
10. To test, type `config edit` into the startpage and paste in your `settings.json`. Save and press `esc`, observe that your settings have now changes and the startpage reflects your settings.
  - However, note that this way of editing cannot update assets. I've put a wallpaper in the `assets` folder and am trying to access it in my config, but this wallpaper won't display until I've fixed this.

### Expected Behavior
I am able to view a startpage that reflects the state of my `settings.json` file, either dynamically or by restarting the server or rebuilding.

### Specs

- OS: Arch Linux, no version because rolling release, I guess my kernel is `6.8.1-arch1-1` if that matters?
- Browser: Firefox 124.0

**Comment by excalith** [comment]
Hey @revsuine, thanks so much for the detailed explanation, it really helps to understand the way you prefer using, besides from diagnosing the issue.

The way you prefer using is the simplest way, which aims to inexperienced users to be able to edit their configs on the fly, or even deploy to their hosts without needing the source (in this case, `settings.json` file). For this reason, the only viable configuration option is to use `config edit` command to change your settings, or use `config import` command to import your configuration (ie. from your dotfiles) directly into the app.

I may look further into hot reloading stuff an option to bypass this functionality later, sounds a nice feature for more advanced users, but can't promise for this right now.

The preferred usage is through containers where you can mount those settings files into your dotfiles folder, so you will have absolute control over the settings, assets etc. directly synced with your dotfiles repository.

Please let me know what you think about it, and feel free to close the issue if it answers your problem.

PS: I *used* arch btw :D The requirement was there to see the OS only, just in case the Docker version creeps out on different platforms.

**Comment by revsuine** [comment]
I see, I didn't know you weren't meant to edit settings.json directly. So there is no way to use my own wallpaper from a local file?

**Reply by excalith** [reply]
Sorry, my previous message might be unclear. There are ways to fully customize it and you can find more info about it on this [wiki page](https://github.com/excalith/excalith-start-page/wiki/Getting-Started)

- If you use the prebuilt online version, you can only give an URL to load the file, not from local drive due to security restrictions
- If you have the source code (this is your case) you can add your assets (wallpaper, icon etc) into `data/assets` and edit them in your `settings.json` file
- If you use Docker with mounts, you will have the data folder exposed so whenever you update something (asset or settings), it will be also read from those sources directly

Hope this helps!

**Reply by revsuine** [reply]
> If you have the source code (this is your case) you can add your assets (wallpaper, icon etc) into data/assets and edit them in your settings.json file

This is what I've done, but my assets aren't displaying on my start page; it's just a solid background for me even though I've set the background to an image in `data/assets`. Again I'm self hosting a fork. How do I get it to see the image file? Do I need an absolute filepath or?

**Reply by excalith** [reply]
Hmm, it should read the path relative to `data` folder. This folder is also exposed for mounting on `docker` containers to edit assets on the fly. Since you have the source, you can add files into that folder and give the relative path to the asset as well.

For example;
1. Put a wallpaper into `data/assets` folder
2. Consider the filename and extension (ie. `wallpaper.png`)
3. Update `wallpaper.url` into `assets/wallpaper.png` in your `settings.json` (It reads from `data` folder, you have to put `assets` folder before file)
4. Since the way you use requires browsers local storage, you need to use `config reset` to reset settings to your settings file.

You could omit the assets folder by placing file into `data` folder directly and put the filename.extension to settings, but this way keeps things more organized.

I know it sounds terrible to resetting back to settings file, but this usage mostly required for editing through `config edit` command, especially useful through using mobile devices such as tablets without source editing requirements.

Hope this solves your question and please do not forget to mark as answer if it does.
Cheers!

**Reply by revsuine** [reply]
Thanks, but that doesn't seem to change anything for me; it's still not loading my wallpaper after tying `config reset`. The path I put was `assets/pure-julia.png`, which is the filepath relative to the `data` directory. Any ideas?
